### PR TITLE
Git-sync atomic initial pull

### PR DESCRIPTION
**Pull request by gregory-lyons**
@huggsboson @ghodss

This implementation will probably change depending on feedback from people on the Kubernetes team.

Just wanted to take an initial stab at the logic and see what you guys think. Will need to update docs as well.

### CODE REVIEW

File: git-sync/atomic.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+       {
+           name: "git-data",
+           emptyDir: {},
+       },
+   ],
+   ...
+}
+```
+
+## Code Implementation
```
Comment (rw103) [inline git-sync/atomic.md]: Looks like the paragraphs and the pseudocode are explaining the same thing? If so, I'd keep the paragraphs high-level (e.g. don't refer to local variable names) and let the pseudocode cover the nitty-gritty.

File: git-sync/atomic.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+For example, a webserver container might run alongside the git-sync sidecar and serve content from the Git repo. The webserver would not want to begin serving the content in its target directory until the data is fully pulled into the volume by git-sync.  The process involves cloning and pulling initially to a temporary directory, setting a symbolic link, and then renaming the symbolic link to the desired target name. We can ensure that the target directory name will only appear to the webserver in the volume filesystem once the pull is complete.
+
+## Usage
+
+In the configuration of the git-sync container, set the optional GIT_SYNC_ATOMIC environment variable to true. The GIT_SYNC_SUBDIR environment variable must also be set to a non-empty string (otherwise an error will result). This subdirectory will be created within the shared volume's root directory, specified by GIT_SYNC_DEST (which defaults to "/git" if unspecified). If specified, GIT_SYNC_DEST should match the mountPath for git-sync's mounted emptyDir volume.
```
Comment (rw103) [inline git-sync/atomic.md]: The Kube folks may have their own opinions, but letting people set `GIT_SYNC_SUBDIR` (as opposed to just picking something for them) seems like it adds complexity but not a lot of value. Putting the Git repo in one dir if atomic is off and a subdir of that if atomic is on may also be a bit for users to remember.

File: git-sync/main.go [hunk 3]
```diff
@@ -168,12 +184,26 @@ func syncRepo(repo, dest, branch, rev string, depth int) error {
...
 		// set file permissions
-		_, err = runCommand("chmod", "", []string{"-R", string(*flChmod), dest})
+		_, err = runCommand("chmod", "", []string{"-R", string(*flChmod), pullTarget})
 		if err != nil {
 			return err
 		}
 	}
 
+	if atomic && strings.Compare(pullTarget, path.Join(dest, subdir)) != 0 {
+		args := []string{"-snf", pullTarget, "data-link"}
```
Comment (rw103) [inline git-sync/main.go]: Is there a benefit to declaring the args right before using them? That var doesn't seem to be used anywhere else.

File: git-sync/main.go [hunk 4]
```diff
@@ -168,12 +184,26 @@ func syncRepo(repo, dest, branch, rev string, depth int) error {
...
+		_, err := runCommand("ln", dest, args)
+		if err != nil {
+			return fmt.Errorf("error creating symbolic link: %v", err)
+		}
+
```
Comment (rw103) [inline git-sync/main.go]: Add a comment explaining why we do an `ln` followed by an `mv` (and not just a single `ln`)?

File: git-sync/atomic.md (hunk 2 above)
Comment (gregory-lyons) [inline git-sync/atomic.md]: The repo has to end up in a subdirectory for atomic to work. Do you think that we should just put the repo in a default subdirectory ($GIT_SYNC_DEST/data) even if atomic is off? 

It seems intuitive to me that $GIT_SYNC_DEST is supposed to be the target for where your content ends up. I figured we should keep that behavior if possible in the normal case (atomic is off by default), and that if someone wanted to turn on atomic, they could deal with the weird subdir requirement (and we force them to explicitly specify the subdir to remind them where the content is going to end up).

If we want to do it the other way, we may want to rethink the naming of GIT_SYNC_DEST and indicate that it points to the root of the shared volume, but not the repo content itself. 

@huggsboson

File: git-sync/main.go (hunk 3 above)
Comment (gregory-lyons) [inline git-sync/main.go]: I used the existing runCommand function for consistency's sake, and it takes the args as a []string.

Comment (gregory-lyons) [inline git-sync/main.go]: Although I could do it in-line like:

```
runCommand("ln", dest, []string{"-snf", pullTarget, "data-link"})
```

File: git-sync/main.go (hunk 4 above)
Comment (gregory-lyons) [inline git-sync/main.go]: I was following pseudocode from thockin that seemed to imply "ln" was not atomic. After a little bit of research, it seems that "ln -f" replaces existing symlinks in a non-atomic way, where "mv" is guaranteed atomic. However, if there is no existing symlink at the path, the single "ln" appears to be atomic (from my primitive test). We will want to do ln + mv for refresh pulls though. Not really sure if I should switch it or not.

**Pull request merged**